### DISCUSSION TITLE
posts/fiber_in_cpp_understanding_the_basics/

### DISCUSSION THREAD

**Discussion by giscus** [JiayinCao/AGraphicsGuyNotes#16] [Giscus-Comments]
# posts/fiber_in_cpp_understanding_the_basics/

Fiber, a less known concept compared with coroutine, is a pretty powerful addition to cooperative multitasking. As a graphics programmer in game industry, I totally appreciate the great flexibility that fiber brings on the table. As a matter of fact, I feel the tech is a little bit underappreciated due to the lack of sufficient public materials talking about fibers.
In this blog post, I will put down some of my learnings about fiber basics.

https://agraphicsguynotes.com/posts/fiber_in_cpp_understanding_the_basics/

**Comment by IJzerbaard** [comment]
It is actually not the case that all SIMD registers are caller-save: xmm6 through xmm15 are callee-save. What makes this perhaps confusing is that ymm0 through ymm15 are listed as caller-save, but for ymm6 through ymm15 what that really means is that their *upper halves* are caller-save, while their lower halves are callee-save (so yes, the same register is partly caller-save, partly callee-save, which sounds stupid but it's for backward-compatibility with pre-AVX code).

**Reply by JiayinCao** [reply]
Thanks for pointing this out.

Here on this [wiki page](https://en.wikipedia.org/wiki/X86_calling_conventions#System_V_AMD64_ABI), it does state this
```
If the callee wishes to use registers RBX, RSP, RBP, and R12–R15, it must restore their original values before returning control to the caller. All other registers must be saved by the caller if it wishes to preserve their values
```

I believe the source of this statement comes from this drafted [document](https://www.uclibc.org/docs/psABI-x86_64.pdf) on page 21. And this document has an official version [here](https://gitlab.com/x86-psABIs/x86-64-ABI/-/jobs/artifacts/master/raw/x86-64-ABI/abi.pdf?job=build).  Neither of the documents mention that `xmm*` are callee-saved.

I did point out in the post that this implementation is based on this ABI.

Besides this, looking at an open source [implementation](https://github.com/google/marl/blob/main/src/osfiber_asm_x64.h) of fiber from Google, none of the `xmm*` registers get cached in the fiber context. This does give me a bit of confidence about what I said in my post.

Do you think there is any chance that you are talking about other ABIs? I did see some other ABI mentioned what you pointed out here, for example [here](https://learn.microsoft.com/en-us/cpp/build/x64-software-conventions?view=msvc-170).

I would be happy to fix it if I'm indeed wrong in the post.

**Comment by agsb** [comment]
In Forth a task just need save the pointers of return and data stack, and swap new user area, get pointers and go.

**Comment by olk** [comment]
C++ fibers are proposed in the ISO C++ standardization paper ["P0876R13 -  fiber_context - fibers without scheduler"](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p0876r13.pdf) and implementation for various architectures is provided by [boost.context](https://www.boost.org/doc/libs/1_83_0/libs/context/doc/html/context/ff.html).

**Comment by FreddyOm** [comment]
Thank you for this incredible article! I was looking for some better information about fibers for quite a while.
This is exactly what I was looking for. It already helped me a lot working on my own fiber based job system!

Keep the good work up!

**Comment by very-soft-particles** [comment]
Hi!
I've been working on my own game engine and tried making a fiber job system using some of your ideas and also stuff from the "Fibers, Oh My" post, plus watching the GDC talk about the Naughty Dog engine a bazillion times.. and I got it working!... more or less
I have a simple vulkan forward renderer I tried to initialize in an init job, everything goes fine until it tries initializing the window, specifically at CreateWindow... I assumed this was because of blocking calls the OS is making in that function, so I moved it into a short initialization phase before the fibers start doing anything and kept the IO loop inside the main thread.. next it crashed at vkCreateInstance but even makes it as far into the function as making the first allocation.. I checked the vulkan spec and it implies that the VkInstance object in vkCreateInstance is internally synchronized which maybe means it also has a blocking call?
I guess I just want to flag that for people trying to do this kind of thing as well.
Also, I could definitely be mistaken, and my fibers are crashing for another unknown reason, but this seems to be what's happening.
Do you have any ideas for work arounds?
In the Naughty Dog talk, the presenter mentions the adaptive_mutex but I don't know how to replace what is potentially a driver blocking mutex (perhaps a Critical Section) with my own synchronization structures and feel that would be risky. 
My workaround for now is having each thread execute all of these "interally synchronized" tasks before launching into the fibers, and then once the fibers yield back into the thread context, it can call the appropriate destructors for the initialized objects within the same thread context.
I haven't sufficiently tested this yet, but I don't see why it won't work seeing as everything can initialize properly in the main thread anyways, it's just the fibers that crash.
I just want to know if there's something I'm missing about all of this!
I will post some code, but I'm really bad about using github. I'm not currently working in programming. Just doing it for the love! 
Thanks for the amazing post!

**Comment by kedouduan** [comment]
Thank you for this helpful post!
